Document CAST and TRY_CAST expressions

kagamiori opened the pull request:
Summary: Add documentation about cast and try_cast functions, as well as a list of supported conversions.

Differential Revision: D47798592

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,269 @@
+
+    Like :func:`cast`, but returns null if the cast fails. ``try_cast(x AS type)``
+    is different from ``try(cast(x AS type))`` in that when there is an error
+    during the evaluation of ``x``, ``try_cast`` throws, while ``try(cast())``
```
This description seems incorrect. It reads that try_cast hides exceptions that happen when evaluating 'x'. However, try_cast should not affect evaluation of 'x', rather hide exceptions that occur while casting x. CC: @laithsakka

kagamiori replied:
I think I meant the same thing. What about I rewrite it to ``try_cast`` throws when there is an error during the evaluation of ``x``, while ``try(cast())`` suppresses the error and turns it to NULL in the result vector?

mbasmanova replied:
I see. I'm afraid other readers may be as confused as I was. Maybe provide an example?

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,269 @@
+    Like :func:`cast`, but returns null if the cast fails. ``try_cast(x AS type)``
+    is different from ``try(cast(x AS type))`` in that when there is an error
+    during the evaluation of ``x``, ``try_cast`` throws, while ``try(cast())``
+    suppresses this error and produces a NULL result at the current row row.
```
typo: 'current row row.'

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,269 @@
+Supported Conversions
+---------------------
+
+The supported conversions in Velox are listed below. Conversions of ARRAY, MAP,
```
drop "in Velox" as it is implied

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,269 @@
+
+The supported conversions in Velox are listed below. Conversions of ARRAY, MAP,
+and ROW types are supported if the conversion of their element types are
+supported. In addition to this list, supported conversions to/from JSON are
```
"In addition to this list," -> In addition,"

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,269 @@
+
+   * - Return types
+     - Argument types
+     - Truncate support
```
"Truncate support" appears for the first time here. Would you explain what it is earlier in the doc?

kagamiori replied:
Unfortunately no. I'm not exactly sure how to describe it since it's controlled by query configuration. Do you know if it is equivalent to Presto truncate() function? I can remove this column for now and add it later when I figure it out.

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,269 @@
+   * - tinyint
+     - tinyint
+     - N
+   * - tinyint
```
This table is long and a bit hard to read. I wonder if some different layout will work better. Maybe nxn matrix with from types in rows and to types in columns? Maybe a few matrices?

mbasmanova reviewed (commented):
@kagamiori Thank you for adding this documentation.

kagamiori commented:
Updated based on comments. Here is what it looks like now.
![Screenshot 2023-07-27 at 2 56 02 PM](https://github.com/facebookincubator/velox/assets/16028358/ccd8f07b-9a89-42c5-b177-7716b24ff189)

mbasmanova commented:
@kagamiori This is nice. Is this matrix symmetrical, e.g. if cast a -> b is supported, then cast b -> a is also supported? If so, we can remove "to" and "from" from the column and row titles .

kagamiori commented:
> @kagamiori This is nice. Is this matrix symmetrical, e.g. if cast a -> b is supported, then cast b -> a is also supported? If so, we can remove "to" and "from" from the column and row titles .

The matrix is not symmetrical. I removed "to" and "from" from the columns and rows and added an explanation in the text.

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,266 @@
+
+::
+
+  SELECT cast(12345.12 as bigint); -- 12345
```
It might be helpful to organize this by from -> to cast, e.g. have a separate section for each cast with both valid and invalid examples as well as English description of valid inputs.

kagamiori commented:
Screenshot of the updated page:
![Screenshot 2023-07-31 at 6 22 34 PM](https://github.com/facebookincubator/velox/assets/16028358/bf0318ec-5860-4e26-bb1a-9674d501406e)

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,521 @@
+  SELECT cast(123.45 as real); -- 123.45
+
+There are two cases where `Velox behaves differently from Presto`_ when casting
+to real from a value beyond real's limit. We will fix them to follow Presto's
```
Would you include links to corresponding GitHub issues? Use 

```
:issue:`1234` syntax.
```

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,521 @@
+^^^^^^^^^^^^
+
+Casting from a string to timestamp is allowed if the string represents a
+timestamp in the valid format. Casting from invalid input values throws.
```
> in the valid format

It would be nice to clarify what is "valid format".

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,521 @@
+^^^^^^^^^^^^
+
+Casting from a string to date is allowed if the string represents a date in the
+valid format. Casting from invalid input values throws.
```
ditto

mbasmanova commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -0,0 +1,521 @@
+From timestamp
+^^^^^^^^^^^^^^
+
+Casting from timestamp to date is allowed.
```
Would you clarify what happens with HH:MM:SS part of the timestamp?

mbasmanova reviewed (approved):
@kagamiori Looks great. Thank you for explaining CAST semantics clearly. A couple of nits below.
